### ISSUE TITLE
New client thread pool has shut down unexpectedly

### ISSUE THREAD

**Issue by idzivinskyi** [akka/akka-http#1612]
With `new` thread pool I got a dozen error messages: 

```
connection pool for PoolGateway(hcps = HostConnectionPoolSetup(******.com,80,ConnectionPoolSetup(ConnectionPoolSettings(20,4,5,128,1,30 seconds,ClientConnectionSettings(Some(User-Agent: akka-http/10.0.11),10 seconds,1 minute,512,None,<function0>,List(),ParserSettings(16384,16,64,64,8192,64,8388608,256,1048576,Strict,RFC6265,true,Full,Error,Map(If-Range -> 0, If-Modified-Since -> 0, If-Unmodified-Since -> 0, default -> 12, Content-MD5 -> 0, Date -> 0, If-Match -> 0, If-None-Match -> 0, User-Agent -> 32),false,true,akka.stream.impl.ConstantFun$$$Lambda$286/1920618675@1b3628c0,akka.stream.impl.ConstantFun$$$Lambda$286/1920618675@1b3628c0,akka.stream.impl.ConstantFun$$$Lambda$287/213044826@1827b4e5),None),New,1 second,TCPTransport),akka.http.scaladsl.HttpConnectionContext$@18e0b4d5,akka.event.MarkerLoggingAdapter@1d9b91a5))) has shut down unexpectedly
```

I have four different queues which `cachedHostConnectionPool`:

```
    val pool = if (uri.effectivePort == 443) {
      Http().cachedHostConnectionPoolHttps[Promise[HttpResponse]](uri.authority.host.address, uri.effectivePort)
    } else {
      Http().cachedHostConnectionPool[Promise[HttpResponse]](uri.authority.host.address, uri.effectivePort)
    }

    Source.queue[(HttpRequest, Promise[HttpResponse])](queueSize, OverflowStrategy.dropNew)
      .via(pool)
      .toMat(Sink.foreach {
        case ((Success(resp), p)) => p.success(resp)
        case ((Failure(e), p)) => p.failure(e)
      })(Keep.left)
      .run
```

What could be the reason of it?

**Comment by jrudolph** [comment]
Sounds like a bug, could you see if there are more log messages on DEBUG level that would point towards what the issue would be?

**Comment by idzivinskyi** [comment]
Yes, no problem. I will come back here when I get results.

**Comment by idzivinskyi** [comment]
@jrudolph 

```
Error in stage [akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage@58fc884d]: Cannot open connection when slot still has an open connection
```

```
connection pool for PoolGateway(hcps = HostConnectionPoolSetup(***.com,80,ConnectionPoolSetup(ConnectionPoolSettings(4,0,5,32,1,30 seconds,ClientConnectionSettings(Some(User-Agent: akka-http/10.0.11),10 seconds,1 minute,512,None,<function0>,List(),ParserSettings(16384,16,64,64,8192,64,8388608,256,1048576,Strict,RFC6265,true,Full,Error,Map(If-Range -> 0, If-Modified-Since -> 0, If-Unmodified-Since -> 0, default -> 12, Content-MD5 -> 0, Date -> 0, If-Match -> 0, If-None-Match -> 0, User-Agent -> 32),false,true,akka.stream.impl.ConstantFun$$$Lambda$289/1244297277@31d39bd1,akka.stream.impl.ConstantFun$$$Lambda$289/1244297277@31d39bd1,akka.stream.impl.ConstantFun$$$Lambda$290/1839014982@6f93939e),None),New,1 second,TCPTransport),akka.http.scaladsl.HttpConnectionContext$@353d5229,akka.event.MarkerLoggingAdapter@3c16c72d))) has shut down unexpected
```

```
Pool stopped
```

```
Host connection pool has shut down with error Cannot open connection when slot still has an open connection
```

```
(Re-)starting host connection pool to ***.com:80
```

```
InputBuffer (max-open-requests = 32) now filled with 1 request after enqueuing POST /some/path Strict(788 bytes)
```

**Comment by idzivinskyi** [comment]
@jrudolph here code to reproduce:

```
import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.client.RequestBuilding.Get
import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
import akka.http.scaladsl.server.Route
import akka.http.scaladsl.settings.{ConnectionPoolSettings, PoolImplementation}
import akka.stream.scaladsl.{Keep, Sink, Source}
import akka.stream.{ActorMaterializer, OverflowStrategy, QueueOfferResult}
import com.grammarly.facade.routes.RouteModule.DefaultRoute
import com.typesafe.scalalogging.StrictLogging

import scala.concurrent.{ExecutionContext, Future, Promise}
import scala.util.{Failure, Success}

object TestPool extends App with DefaultRoute with StrictLogging {

  implicit val system: ActorSystem = ActorSystem("system")

  implicit val executionContext: ExecutionContext = system.dispatcher

  implicit val materializer: ActorMaterializer = ActorMaterializer()

  private val port = 8080

  private val settings = ConnectionPoolSettings(system).withPoolImplementation(PoolImplementation.New)
//    private val settings = ConnectionPoolSettings(system).withPoolImplementation(PoolImplementation.Legacy)

  val queue = Source.queue[(HttpRequest, Promise[HttpResponse])](20, OverflowStrategy.backpressure)
    .via(Http().cachedHostConnectionPool[Promise[HttpResponse]]("localhost", port, settings))
    .toMat(Sink.foreach {
      case ((Success(resp), p)) => p.success(resp)
      case ((Failure(e), p)) => p.failure(e)
    })(Keep.left)
    .run


  val route = path("ping") {
    get {
      logger.info("ask_ping")
      val promise = Promise[HttpResponse]

      val future: Future[HttpResponse] = queue.offer(Get("http://localhost:8080/ask_ping") -> promise).flatMap {
        case QueueOfferResult.Enqueued =>
          promise.future
        case QueueOfferResult.Dropped =>
          Future.failed(new RuntimeException("Dropped"))
        case QueueOfferResult.Failure(ex) =>
          Future.failed(ex)
        case QueueOfferResult.QueueClosed =>
          Future.failed(new RuntimeException("QueueClosed"))
      }

      onSuccess(future) { res =>
        complete(res)
      }
    }
  } ~ path("ask_ping") {
    get {
      logger.info("return_pong")
      complete("pong")
    }
  }

  val server: Future[Http.ServerBinding] = Http().bindAndHandle(Route.handlerFlow(route), "localhost", port)

  server.foreach { b =>
    logger.info(s"Server started on ${b.localAddress.toString}")
  }

  (1 to 100).foreach { _ =>
    queue.offer(Get("http://localhost:8080/ping") -> Promise[HttpResponse])
    Thread.sleep(50)
  }
}
```

**Comment by idzivinskyi** [comment]
@jrudolph Here code to reproduce:

```
import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.client.RequestBuilding.Get
import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
import akka.http.scaladsl.server.{Directives, Route}
import akka.http.scaladsl.settings.{ConnectionPoolSettings, PoolImplementation}
import akka.stream.scaladsl.{Keep, Sink, Source}
import akka.stream.{ActorMaterializer, OverflowStrategy, QueueOfferResult}
import com.typesafe.scalalogging.StrictLogging

import scala.concurrent.{ExecutionContext, Future, Promise}
import scala.util.{Failure, Success}

object TestPool extends App with Directives with StrictLogging {

  implicit val system: ActorSystem = ActorSystem("system")

  implicit val executionContext: ExecutionContext = system.dispatcher

  implicit val materializer: ActorMaterializer = ActorMaterializer()

  private val port = 8080

  private val settings = ConnectionPoolSettings(system).withPoolImplementation(PoolImplementation.New)
//    private val settings = ConnectionPoolSettings(system).withPoolImplementation(PoolImplementation.Legacy)

  val queue = Source.queue[(HttpRequest, Promise[HttpResponse])](20, OverflowStrategy.backpressure)
    .via(Http().cachedHostConnectionPool[Promise[HttpResponse]]("localhost", port, settings))
    .toMat(Sink.foreach {
      case ((Success(resp), p)) => p.success(resp)
      case ((Failure(e), p)) => p.failure(e)
    })(Keep.left)
    .run


  val route = path("ping") {
    get {
      logger.info("ask_ping")
      val promise = Promise[HttpResponse]

      val future: Future[HttpResponse] = queue.offer(Get("http://localhost:8080/ask_ping") -> promise).flatMap {
        case QueueOfferResult.Enqueued =>
          promise.future
        case QueueOfferResult.Dropped =>
          Future.failed(new RuntimeException("Dropped"))
        case QueueOfferResult.Failure(ex) =>
          Future.failed(ex)
        case QueueOfferResult.QueueClosed =>
          Future.failed(new RuntimeException("QueueClosed"))
      }

      onSuccess(future) { res =>
        complete(res)
      }
    }
  } ~ path("ask_ping") {
    get {
      logger.info("return_pong")
      complete("pong")
    }
  }

  val server: Future[Http.ServerBinding] = Http().bindAndHandle(Route.handlerFlow(route), "localhost", port)

  server.foreach { b =>
    logger.info(s"Server started on ${b.localAddress.toString}")
  }

  (1 to 100).foreach { _ =>
    queue.offer(Get("http://localhost:8080/ping") -> Promise[HttpResponse])
    Thread.sleep(50)
  }
}
```

**Comment by jrudolph** [comment]
We expect this to be fixed with #1655.
